lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs

```diff
@@ -723,43 +738,79 @@ monitorMetadata gcStatus tr sp db@(DBLayer{..}) = do
         _ -> pure NoSmashConfigured
 
     if  | health == Available || health == NoSmashConfigured -> do
-            let fetcher fetchStrategies = fetchFromRemote trFetch fetchStrategies manager
-                loop getPoolMetadata = forever $ do
-                    (refs, successes) <- getPoolMetadata
-                    when (null refs || null successes) $ do
-                        traceWith tr $ MsgFetchTakeBreak blockFrequency
-                        threadDelay blockFrequency
-
             case poolMetadataSource settings of
                 FetchNone -> do
                     STM.atomically $ writeTVar gcStatus NotApplicable
-                    loop (pure ([], [])) -- TODO: exit loop?
+
                 FetchDirect -> do
                     STM.atomically $ writeTVar gcStatus NotApplicable
-                    loop (fetchThem $ fetcher [identityUrlBuilder])
+                    queue <- STM.atomically $ newTBQueue queueSize
+                    concurrently_
+                        (enqueueMetadata queue)
+                        (dequeueMetadata manager [identityUrlBuilder] queue)
+
                 FetchSMASH (unSmashServer -> uri) -> do
                     STM.atomically $ writeTVar gcStatus NotStarted
                     let getDelistedPools =
                             fetchDelistedPools trFetch uri manager
                     tid <- forkFinally
                         (gcDelistedPools gcStatus tr db getDelistedPools)
                         onExit
+                    queue <- STM.atomically $ newTBQueue queueSize
                     flip finally (killThread tid) $
-                        loop (fetchThem $ fetcher [registryUrlBuilder uri])
-        | otherwise -> traceWith tr MsgSMASHUnreachable
+                        concurrently_
+                            (enqueueMetadata queue)
+                            (dequeueMetadata manager [registryUrlBuilder uri] queue)
 
+        | otherwise ->
+            traceWith tr MsgSMASHUnreachable
   where
     trFetch = contramap MsgFetchPoolMetadata tr
-    fetchThem fetchMetadata = do
+
+    queueSize :: Natural
+    queueSize = 100
+
+    -- Must be:
+    --     - > 1
+    --     - <= queueSize
+    -- Should be:
+    --     - <= queueSize / 2
+    numInFlight :: Int
+    numInFlight = 20
+
+    enqueueMetadata
+        :: TBQueue (PoolId, StakePoolMetadataUrl, StakePoolMetadataHash)
+        -> IO ()
+    enqueueMetadata queue = forever $ do
         refs <- atomically (unfetchedPoolMetadataRefs 100)
```

Comment:
Perhaps set this number of refs to be smaller than queueSize - maybe the same as numInFlight.

Reply:
Whoops, I should've removed the magic number here and use `queueSize` instead. Thing is, I don't want the queue to be empty if there's something to fetch. So ideally, this should be > numInFlight. The bounded queue will block on writes anyway if the queue is full.